karapace/schema_reader.py

```diff
@@ -250,109 +262,131 @@ def handle_messages(self):
+        else:
+            self.log.info("Deleting subject: %r, value: %r", subject, value)
+            updated_schemas = {
+                key: self._delete_schema_below_version(schema, value["version"])
+                for key, schema in self.subjects[subject]["schemas"].items()
+            }
+            self.subjects[value["subject"]]["schemas"] = updated_schemas
+
+    def _handle_msg_schema_hard_delete(self, key: dict) -> None:
+        subject, version = key["subject"], key["version"]
+
+        if subject not in self.subjects:
+            self.log.error("Hard delete: Subject %s did not exist, should have", subject)
+        elif version not in self.subjects[subject]["schemas"]:
+            self.log.error("Hard delete: Version %d for subject %s did not exist, should have", version, subject)
+        else:
+            self.log.info("Hard delete: subject: %r version: %r", subject, version)
+            self.subjects[subject]["schemas"].pop(version, None)
+
+    def _handle_msg_schema(self, key: dict, value: Optional[dict]) -> None:
+        if not value:
+            self._handle_msg_schema_hard_delete(key)
+            return
+
+        schema_type = value.get("schemaType", "AVRO")
+        schema_str = value["schema"]
+        schema_subject = value["subject"]
+        schema_id = value["id"]
+        schema_version = value["version"]
+        schema_deleted = value.get("deleted", False)
+
+        try:
+            schema_type_parsed = SchemaType(schema_type)
+        except ValueError:
+            self.log.error("Invalid schema type: %s", schema_type)
+            return
+
+        # Protobuf doesn't use JSON
+        if schema_type_parsed in [SchemaType.AVRO, SchemaType.JSONSCHEMA]:
+            try:
+                ujson.loads(schema_str)
+            except ValueError:
+                self.log.error("Invalid json: %s", value["schema"])
+                return
+
+        typed_schema = TypedSchema(schema_type=schema_type_parsed, schema_str=schema_str)
+        self.log.debug("Got typed schema %r", typed_schema)
+
+        if schema_subject not in self.subjects:
+            self.log.info("Adding first version of subject: %r with no schemas", schema_subject)
+            self.subjects[schema_subject] = {"schemas": {}}
+
+        subjects_schemas = self.subjects[schema_subject]["schemas"]
+
+        if schema_version in subjects_schemas:
+            self.log.info("Updating entry for subject: %r, value: %r", schema_subject, value)
+        else:
+            self.log.info("Adding new version of subject: %r, value: %r", schema_subject, value)
+
+        subjects_schemas[schema_version] = {
+            "schema": typed_schema,
+            "version": schema_version,
+            "id": schema_id,
+            "deleted": schema_deleted,
+        }
+        with self.id_lock:
+            self.schemas[schema_id] = typed_schema
+            self.global_schema_id = max(self.global_schema_id, schema_id)
```

Comment:
This is the bugfix for the `global_schema_id` mentioned above: https://github.com/aiven/karapace/pull/375/files#r839829337

The fix is simple, always update the `global_schema_id`.

Reply:
Could you elaborate on this a bit? Say our `global_schema_id` counter is at `5` and then we register another subject that will reuse an older schema that we have with `schema_id: 1`. Wouldn't updating to that eventually lead to duplicate id's or did you have some other update in mind?